Lilac: Remote gradebook & Canvas cherry-picks

umarmughal824 opened the pull request:
fixes #255 

Cherry-pick the following commits into one commit and test that it works on the mitx/lilac branch.

(most recent at the top)
608828d
e630376
e52d609
b331778
697ce49
1374507
63ceb85
ca3a1a6

I don't know how to test it but after making all these changes, my devstack is still functional. I had faced some import issues that I have fixed in the last commit to make it functional.

#### Screenshots

<img width="1439" alt="Screenshot 2021-04-15 at 17 10 47" src="https://user-images.githubusercontent.com/4043989/114866957-a5e5db00-9e0d-11eb-8026-b370df0aed11.png">

<img width="1440" alt="Screenshot 2021-04-15 at 17 11 01" src="https://user-images.githubusercontent.com/4043989/114866966-a7af9e80-9e0d-11eb-83e1-71f14816350b.png">

gsidebo committed:
Reimplemented legacy dashboard enrollment data functionality (60a6ddc7c1)

asadiqbal08 committed:
Canvas Commits (#223)

* Sync canvas enrollments

* Added instructor dashboard button to push edX grades to Canvas

* Fixed canvas grade syncing JS

* Changed 'edX' reference to 'MITx'

* Use EDIT_COURSE_ACCESS permission instead of is_staff

* Use OVERRIDE_GRADES rule instead

* Use instructor_tasks for canvas work (#183)

* Add per_page to paginated requests

* Static method

* Fix email lowercase mismatch

* Use instructor_tasks to handle canvas work

* Fix transaction error

* Decorator needs to be at the top

* Attempt to fix polling of tasks

* More task bug fixes

* Fix typo

* Fix course key bug

* Update message for push edx grades command

* Use course_key so sync_canvas_enrollments uses the same message

* Format the submitted time

* Don't return output to prevent task output max size error (#189)

Co-authored-by: George Schneeloch <gschneel@mit.edu>
Co-authored-by: Gavin Sidebottom <gsidebo@mit.edu>

shaidar committed:
Fix syntax error in import statement

arslanashraf7 committed:
Cherry-Picked: Enrolment commits (#228)

* Adding enrollment restriction for users based on social auth providers

In order to prevent users outside of MIT creating a collaborator account and self-enrolling in courses we need to be able to restrict self-enrollment to users who authenticate using the MIT Kerberos SAML backend. This adds filtering of self-enrollment based on a feature flag that sets the allowed provider ids.

(cherry picked from commit a4c53a1c91a40b54c24a94648ae3c40c9b76d85a)

* Bugfix for collaborator account access check

Co-authored-by: Tobias Macey <tmacey@mit.edu>

blarghmatey committed:
Fixing import path for canvas_integration plugin

blarghmatey committed:
Fix import namespace for relative imprts

edX requires the full namespace for imports so importing e.g. `remote_gradebook.tasks` is no longer valid. Instead it
requires `from lms.djangoapps.remote_gradebook import tasks`.

Commit:
#229 change submission type and published state on canvas (#232)

- Change submission type to "None"
- Change published state to "False"

blarghmatey committed:
Fixing encoding of header row for posting to remote gradebook

Posting to the remote gradebook via lmod proxy was failing due to a key error when parsing the CSV content. This was due to erroneously encoding a string object. This commit adds a check for the type of the header value to prevent double encoding it.

Commit:
fix the imports

pdpinch commented on lms/djangoapps/courseware/access.py:
```diff
@@ -253,6 +253,21 @@ def _can_enroll_courselike(user, courselike):
+                "ENROLLMENT DENIED in %s because user %s not authenticated by any of %s",
+                course_key, user.username, allowed_providers)
+            return ACCESS_DENIED
+
```
This change seems out of place. I think this functionality should already be part of lilac, but possibly under a different setting name. See https://github.com/edx/edx-platform/pull/24789

umarmughal824 replied:
As per my understadning this seems like allowing particular social providers from the list of available.

But edx#24789 seems something different to only allowing third-party authentication and blocking the normal flow of authentication.

@pdpinch What is your final opinion on that?

pdpinch replied:
@mitodl/devops do we need this setting for the residential deployment? 

It looks like it's set in [salt-ops](https://github.com/mitodl/salt-ops/blob/3aa25b111492f7b8f50e552189127f81a6d9f497/pillar/edx/ansible_vars/residential.sls#L175-L176)

@umarmughal824 if we need it, we should keep it in the cherry-pick -- and we should consider upstreaming it to edx.

pdpinch replied:
If we do need to keep this, perhaps it should be in a separate PR, like we did for koa in https://github.com/mitodl/edx-platform/pull/228

blarghmatey replied:
So, this setting is for allowing us to be able to gate enrollments from touchstone collaborators. There is still the open question about whether collaborator accounts are even a thing any more, so we'll likely need to do some investigation about how to manage that restriction going forward.

pdpinch commented on lms/envs/common.py:
```diff
@@ -262,14 +262,12 @@
+    # This will be matched against the beginning portion of the user's social auth uid
+    'RESTRICT_ENROLL_SOCIAL_PROVIDERS': [],
+    # Give a UI to show a student's submission history in a problem by the
+    # Staff Debug tool.
```
I think this also should be removed, in favor of https://github.com/edx/edx-platform/pull/24789

umarmughal824 replied:
Same comment for it.

blarghmatey replied:
@umarmughal824 we do want to keep this code, but it is orthogonal to the Canvas integration so it would be best if we can add it as a separate PR to keep the commit isolated.

pdpinch reviewed (changes requested):
A couple of minor comments. Otherwise, I think we should merge this and see how it works on QA.

Commit:
address feedback

umarmughal824 commented:
@pdpinch I have addressed the feedback so that's ready for review again

@blarghmatey here is the isolated PR for that https://github.com/mitodl/edx-platform/pull/257

blarghmatey commented:
Looks good to me
